[IA-1446] [SATURN-1225] Jc docker documentation

jdcanas opened the pull request:
This leverages the new build scripts in the terra-docker repo, which automatically produce json documentation, along with a file that contains the current versions. This removes the need to PR changes for leo docker versions.

my linter went a little crazy. I reapplied and fixed some issues, but I think the rest of its changes are arguably improvements. 

https://broadworkbench.atlassian.net/browse/IA-1446

Commit:
changes working with json docker data from leo

Commit:
adjusting size of dropdown slightly

Commit:
indentation and moving imagedepviewer to its own component

Commit:
fixing lint problems

Commit:
more lint fixes

Commit:
more lint fixes

panentheos reviewed (commented):
Thanks for the PR. The extraneous formatting changes make it hard to focus on the actual changes. Can you please back out any formatting changes that aren't specifically enforced by the eslint rules? Thanks.

Commit:
changing display of generic tools

jdcanas commented:
> Thanks for the PR. The extraneous formatting changes make it hard to focus on the actual changes. Can you please back out any formatting changes that aren't specifically enforced by the eslint rules? Thanks.

I reverted several (it was a lot worse, vscode is currently bugged as far as I can tell and there is no way to stop/configure it from mangling entire javascript files on save. I had to switch to intellij...). I would like to keep the remaining. I think it's arguably more readable in those cases, and not worth the time to manually revert unless there are strong feelings about the original formatting.

Commit:
update comment

Commit:
simplify constructor

Commit:
wip pr feedback

ehanna4 committed:
Few Mob Fixes

- Lint fixes
- Show dropdown even for single language option (?)
- Add conditional for imageDepViewer rendering

Commit:
addressing additional PR feedback

jdcanas commented on src/components/NewClusterModal.js:
```diff
@@ -139,21 +108,81 @@ export const NewClusterModal = withModalDrawer({ width: 675 })(class NewClusterM
       profiles
     )
+
+    this.fetchMasterVersionFile()
```
move to componentDidMount()

jdcanas commented on src/libs/ajax.js:
```diff
@@ -728,6 +728,17 @@ const Workspaces = signal => ({
 const Buckets = signal => ({
+
+  getFile: async (bucket, object) => {
+    const request = { mode: 'cors' }
```
re-visit

jdcanas commented on src/components/NewClusterModal.js:
```diff
@@ -139,21 +108,81 @@ export const NewClusterModal = withModalDrawer({ width: 675 })(class NewClusterM
 
+  async fetchMasterVersionFile() {
+    const defaultImageId = 'leonardo-jupyter-dev'
+    const file = await Ajax().Buckets.getFile('terra-docker-image-documentation', 'terra-docker-versions.json')
```
getPreview function works?

jdcanas commented on src/components/NewClusterModal.js:
```diff
@@ -139,21 +108,81 @@ export const NewClusterModal = withModalDrawer({ width: 675 })(class NewClusterM
+  async fetchMasterVersionFile() {
+    const defaultImageId = 'leonardo-jupyter-dev'
+    const file = await Ajax().Buckets.getFile('terra-docker-image-documentation', 'terra-docker-versions.json')
+    this.setState({ leoImages: file, selectedLeoImage: file.filter(doc => doc.id === defaultImageId)[0].image })
```
safe, ensure it carries image

jdcanas commented on src/components/NewClusterModal.js:
```diff
@@ -139,21 +108,81 @@ export const NewClusterModal = withModalDrawer({ width: 675 })(class NewClusterM
+    const imageConfig = _.find({ id: defaultImageId }, file)
+    const { packages } = imageConfig
+
+    this.fetchImageDocumentation(packages)
```
consider moving to parent component

jdcanas commented on src/components/NewClusterModal.js:
```diff
@@ -139,21 +108,81 @@ export const NewClusterModal = withModalDrawer({ width: 675 })(class NewClusterM
+    const bucket = splitPath[splitPath.length - 2]
+
+    const file = await Ajax().Buckets.getFile(bucket, object)
+
```
pass link instead splitting, may irrelevant if I leverage existing function

jdcanas commented on src/components/NewClusterModal.js:
```diff
@@ -139,21 +108,81 @@ export const NewClusterModal = withModalDrawer({ width: 675 })(class NewClusterM
+    //transform list of tools into the documentation for the tools section
+    const toolsDoc = {
+      [toolsLabel]: tools
+        .filter(tool => !nonGenericTools.includes(tool.toLowerCase()))
```
use lodash, look for examples _.filter

jdcanas commented on src/components/NewClusterModal.js:
```diff
@@ -139,21 +108,81 @@ export const NewClusterModal = withModalDrawer({ width: 675 })(class NewClusterM
+    if (!_.isEmpty(toolsDoc[toolsLabel])) {
+      finalDoc = _.merge(genericDoc, toolsDoc)
+    }
+
```
explicitly sort?

jdcanas commented on src/components/NewClusterModal.js:
```diff
@@ -139,21 +108,81 @@ export const NewClusterModal = withModalDrawer({ width: 675 })(class NewClusterM
+
+  //takes a packageDoc with n tools, and returns one with at most Python, R, and a generic 'tool' bucket
+  packageDocAdaptor(packageDoc) {
+    const tools = _.keys(packageDoc)
```
look into list of records

jdcanas commented on src/components/NewClusterModal.js:
```diff
@@ -139,21 +108,81 @@ export const NewClusterModal = withModalDrawer({ width: 675 })(class NewClusterM
+      })
+      .reduce((tool1Doc, tool2Doc) => _.merge(tool1Doc, tool2Doc), {})
+
+    let finalDoc = genericDoc
```
ternary operator

jdcanas commented on src/components/NewClusterModal.js:
```diff
@@ -192,18 +222,35 @@ export const NewClusterModal = withModalDrawer({ width: 675 })(class NewClusterM
+      customEnvImage,
+      viewMode,
+      leoImages,
+      packageDoc
```
should extract packageLink not doc

jdcanas commented on src/components/NewClusterModal.js:
```diff
@@ -220,7 +267,7 @@ export const NewClusterModal = withModalDrawer({ width: 675 })(class NewClusterM
           makeEnvSelect(),
           makeImageInfo({ margin: '1rem 0 2rem' }),
-          h(ImageDepViewer, { packages })
+          packageDoc && !_.isEmpty(packageDoc) && h(ImageDepViewer, { packageDoc })
```
should be link not doc

Commit:
initial PR fixes, everything in working state

Commit:
working with fp lodash refactor

Commit:
useEffect working

Commit:
but who lints the linter

jdcanas commented on src/components/ImageDepViewer.js:
```diff
@@ -0,0 +1,109 @@
+      }
+    }
+    inner()
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
without this, the linter complains about `namespace` and `tools` not being included as dependencies, but since the render function interacts with `tools`, doing this results in an infinite re-render loop.

It seems others have had this issue: https://stackoverflow.com/questions/55840294/how-to-fix-missing-dependency-warning-when-using-useeffect-react-hook

Commit:
simplifying code slightly

jdcanas commented on src/components/ImageDepViewer.js:
```diff
@@ -0,0 +1,108 @@
+  }
+
+  useEffect(() => {
+    async function inner() {
```
needs to be done like this

Commit:
simplify logic'

Commit:
error checking

Commit:
fix initial state

Commit:
flatten + map -> flatmap

Commit:
handling loading a bit better

panentheos committed:
cleanup

panentheos commented:
Note: I made some updates to get this toward completion:
* Removed extraneous state
* Initialized form with current cluster values when present
* Simplified packages loading
* Fixed whitespace and stylistic issues

zarsky-broad commented on src/components/ImageDepViewer.js:
```diff
@@ -0,0 +1,69 @@
+    ]),
+    div({ style: { padding: '1rem', marginTop: '1rem', backgroundColor: 'white', borderRadius: 5, flexGrow: 1 } }, [
+      table([
+        thead([
```
Put the `paddingBottom` here on on the `tr`?

zarsky-broad commented on src/components/ImageDepViewer.js:
```diff
@@ -0,0 +1,69 @@
+
+  useEffect(() => {
+    const loadPackages = withErrorReporting('Error loading packages', async () => {
+      const res = await fetchOk(packageLink)
+      const data = await res.json()
```
Why not `.then`?

zarsky-broad commented on src/components/NewClusterModal.js:
```diff
@@ -172,12 +140,18 @@ export const NewClusterModal = withModalDrawer({ width: 675 })(class NewClusterM
+    if (currentClusterDetails) {
+      const { clusterImages, jupyterUserScriptUri } = currentClusterDetails
+      const { dockerImage } = _.find({ tool: 'Jupyter' }, clusterImages)
+      if (_.find({ image: dockerImage }, newLeoImages)) {
```
`_.some`?

zarsky-broad commented on src/components/NewClusterModal.js:
```diff
@@ -210,8 +186,8 @@ export const NewClusterModal = withModalDrawer({ width: 675 })(class NewClusterM
     const makeImageInfo = style => div({ style: { whiteSpace: 'pre', ...style } }, [
-      div({ style: Style.proportionalNumbers }, [`Updated: ${Utils.makeStandardDate(updated)}`]),
-      div([`Version: ${version}`])
+      div({ style: Style.proportionalNumbers }, ['Updated: ', updated ? Utils.makeStandardDate(updated) : null]),
```
Why not just have the `|| null` within the string interpolation?

zarsky-broad commented on src/data/leo-images.js:
```diff
@@ -1,26 +1,2 @@
-]
-
 // distilled from https://github.com/docker/distribution/blob/95daa793b83a21656fe6c13e6d5cf1c3999108c7/reference/regexp.go
 export const imageValidationRegexp = /^[A-Za-z0-9]+[\w./-]+(?::\w[\w.-]+)?(?:@[\w+.-]+:[A-Fa-f0-9]{32,})?$/
```
This is only used in one place, should we inline it and just delete this file?

zarsky-broad commented on src/components/ImageDepViewer.js:
```diff
@@ -0,0 +1,69 @@
+      table([
+        thead([
+          tr([
+            td({ style: { fontWeight: 'bold', paddingRight: '1rem', paddingBottom: '1rem' } }, ['Package']),
```
Maybe a `minWidth` so this shifts around a bit less?

zarsky-broad reviewed (approved):
Looks good, a few comments/thoughts
